discovery-provider/src/tasks/index_metrics.py

```diff
@@ -126,7 +136,155 @@ def refresh_metrics_matviews(db):
+
+    logger.info(f"visited node timestamps: {visited_node_timestamps}")
+    if visited_node_timestamps:
+        redis_set_and_dump(redis, metrics_visited_nodes, json.dumps(visited_node_timestamps))
+
+def get_historical_metrics(node):
+    try:
+        endpoint = f"{node}/v1/metrics/aggregates/historical"
+        logger.info(f"historical metrics request to: {endpoint}")
+        response = requests.get(endpoint, timeout=10)
+        if response.status_code != 200:
+            raise Exception(f"Query to historical metrics endpoint {endpoint} \
+                failed with status code {response.status_code}")
+
+        return response.json()['data']
+    except Exception as e:
+        logger.warning(f"Could not get historical metrics from node {endpoint}")
+        logger.error(e)
+        return None
+
+def update_route_metrics_count(my_metrics, other_metrics):
+    for timestamp, values in other_metrics.items():
+        if timestamp in my_metrics:
+            my_metrics[timestamp] = {
+                'unique': max(values['unique'], my_metrics[timestamp]['unique']),
+                'total': max(values['total'], my_metrics[timestamp]['total'])
+            }
+        else:
+            my_metrics[timestamp] = values
+
+def update_app_metrics_count(my_metrics, other_metrics):
+    for timestamp, values in other_metrics.items():
+        if timestamp in my_metrics:
+            for app, count in values.items():
+                my_metrics[timestamp][app] = max(my_metrics[timestamp][app], count) \
+                    if app in my_metrics[timestamp] else count
+        else:
+            my_metrics[timestamp] = values
+
+def update_historical_metrics(db, daily_route_metrics, monthly_route_metrics, daily_app_metrics, monthly_app_metrics):
+    update_historical_daily_route_metrics(db, daily_route_metrics)
+    update_historical_monthly_route_metrics(db, monthly_route_metrics)
+    update_historical_daily_app_metrics(db, daily_app_metrics)
+    update_historical_monthly_app_metrics(db, monthly_app_metrics)
+
+# get historical monthly metrics and last month's daily metrics to periodically synchronize metrics across all nodes
+def synchronize_all_node_metrics(self, db):
+    daily_route_metrics = {}
+    monthly_route_metrics = {}
+    daily_app_metrics = {}
+    monthly_app_metrics = {}
+    for node in get_all_other_nodes():
+        historical_metrics = get_historical_metrics(node)
+        if historical_metrics:
+            update_route_metrics_count(daily_route_metrics, historical_metrics['routes']['daily'])
+            update_route_metrics_count(monthly_route_metrics, historical_metrics['routes']['monthly'])
+            update_app_metrics_count(daily_app_metrics, historical_metrics['apps']['daily'])
+            update_app_metrics_count(monthly_app_metrics, historical_metrics['apps']['monthly'])
+
+    logger.info("synchronizing historical metrics")
+    logger.info(f"daily historical route metrics to update: {daily_route_metrics}")
```

Comment:
same thing about printing large objects here, do we want this here?

Reply:
these shouldn't be super large though, at least for the route metrics. daily should be about 30 records, each with a unique and total count; monthly would be (12*num_total_years) records. for apps it's harder to predict the number because the number of apps used for a given day/month could be different

but we can surely pull it for now until we see a real need for it